# Exercises I
Create a notebook or python script named advanced_dataframes to do your work in for these exercises.

1. Run python -m pip install pymysql from your terminal to install the mysql client (any folder is fine)
2. cd into your exercises folder for this module and run echo env.py >> .gitignore
3. Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

4. Use your function to obtain a connection to the employees database.

5. Once you have successfully run a query:

- a. Intentionally make a typo in the database url. What kind of error message do you see?

- b. Intentionally make an error in your SQL query. What does the error message look like?

6. Read the employees and titles tables into two separate DataFrames.

7. How many rows and columns do you have in each DataFrame? Is that what you expected?

8. Display the summary statistics for each DataFrame.

9. How many unique titles are in the titles DataFrame?

10. What is the oldest date in the to_date column?

11. What is the most recent date in the to_date column?

In [1]:
import pandas as pd
import numpy as np

### 3. Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [2]:
def get_db_url(db_name):
    from env import host, user, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'
    
# def get_db_url(user,password,host,db_name):
#     return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'


# url = get_db_url(user,password,host,'employees')

### 4. Use your function to obtain a connection to the employees database.



In [3]:
url = get_db_url('employees')

# from env import host, user, password
# url = get_db_url(user,password,host,'employees')

In [4]:
query = 'Select * from employees'

In [5]:
df = pd.read_sql(query, url)
df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

### 5. Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?

b. Intentionally make an error in your SQL query. What does the error message look like?

In [6]:
def get_bad_url(db_name):
    from env import host, user, password
    return f'mysql+pymysql://{user}:{password}@fdas{host}/{db_name}'

bad_url = get_bad_url('employees')
bad_query = 'Select all from employees'

errorA = pd.read_sql(query, bad_url)
errorA

OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'fdas157.230.209.171' (timed out)")
(Background on this error at: http://sqlalche.me/e/14/e3q8)

In [7]:
errorB = pd.read_sql(bad_query, url)
errorB

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'from employees' at line 1")
[SQL: Select all from employees]
(Background on this error at: http://sqlalche.me/e/14/f405)

### 6. Read the employees and titles tables into two separate DataFrames.



In [8]:
query_employees = 'Select * from employees'

In [9]:
employees = pd.read_sql(query_employees, url)
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [10]:
query_titles = 'Select * from titles'

In [11]:
titles = pd.read_sql(query_titles, url)
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

### 7. How many rows and columns do you have in each DataFrame? Is that what you expected?

In [12]:
employees.shape[0]

300024

In [13]:
titles.shape[0]

443308

Employees = 300024 rows, Titles = 443308

### 8.  Display the summary statistics for each DataFrame.



In [14]:
employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [15]:
employees.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [16]:
titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


In [17]:
titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

### 9. How many unique titles are in the titles DataFrame?



In [18]:
len(titles.title.value_counts())

7

### 10. What is the oldest date in the to_date column?



In [19]:
titles.to_date.min()

datetime.date(1985, 3, 1)

### 11. What is the most recent date in the to_date column?



In [20]:
titles_date_query = """SELECT * From titles WHERE to_date != '9999-01-01'"""
titles_recent_date_df = pd.read_sql(titles_date_query, url)
titles_recent_date_df.to_date

0         1995-12-01
1         1996-09-12
2         1996-02-11
3         2000-07-31
4         1990-02-18
             ...    
203179    1992-05-09
203180    1993-10-27
203181    2002-05-13
203182    1992-08-29
203183    1998-12-27
Name: to_date, Length: 203184, dtype: object

In [23]:
titles.to_date.max()
#or filtered to exclude 9999-01-01
titles_recent_date_df.to_date.max()

titles[titles.to_date != titles.to_date.max()].to_date.max()

datetime.date(2002, 8, 1)

# Exercises II

1. Copy the users and roles DataFrames from the examples above.

2. What is the result of using a right join on the DataFrames?

3. What is the result of using an outer join on the DataFrames?

4. What happens if you drop the foreign keys from the DataFrames and try to merge them?

5. Load the mpg dataset from PyDataset.

6. Output and read the documentation for the mpg dataset.

7. How many rows and columns are in the dataset?

8. Check out your column names and perform any cleanup you may want on them.

9. Display the summary statistics for the dataset.

10. How many different manufacturers are there?

11. How many different models are there?

12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

16. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

17. Do automatic or manual cars have better miles per gallon?


